### DISCUSSION TITLE
Help with bulk upsert

### DISCUSSION THREAD

**Discussion by SalmanGits** [prisma/prisma#22688] [Q&A]
### Question

i want to make bulk upsert if user exist then update his score otherwise create

### Information about Prisma Schema, Client Queries and Environment (optional)
```js
 const users = await prisma.user.findMany({});
        let referralCount = {};
        for (let user of users) {
            const uplineId = user.uplineId;

            if (uplineId) {
                referralCount[uplineId] = (referralCount[uplineId] || 0) + 1;
            }
        }
        const purchasedContracts = await prisma.purchasedContract.groupBy({
            by: ["user"],
            _sum: {
                amount: true
            }
        })
        for (let contract of purchasedContracts) {
            const userId = contract.user;
            const amount = contract._sum.amount;
            referralCount[userId] += amount;
        }
        const rewardsData = Object.entries(referralCount).map(([user, score]) => ({
            user,
            score,
        }));
        const upsertedRewards = await prisma.reward.upsert({
            where: { user: { in: Object.keys(referralCount) } },
            create: rewardsData,
            update: {
                data: {
                    score: Object.values(referralCount)
                }
            }
        });
```

**Comment by ludralph** [comment]
Hi @SalmanGits 👋 

Thank you for raising this question.

Prisma Client at the moment does not directly support a bulk upsert through a single command that operates on multiple records at once.

There is an open [feature](https://github.com/prisma/prisma/issues/4134) to support `upsertMany` operation in the Prisma Client API.

As a workaround, you would typically need to iterate over each item you want to upsert and call the upsert method individually for each one. This could be done within a transaction to ensure atomicity.

For example:
```js
const upsertPromises = rewardsData.map((reward) => {
  return prisma.reward.upsert({
    where: { user: reward.user },
    create: reward,
    update: { score: reward.score },
  });
});

const upsertedRewards = await prisma.$transaction(upsertPromises);
```

Alternatively, You can also use raw query to perform the bulk upsert operation.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by SalmanGits** [reply]
yes actually i have  already implemented what you have mentioned here but my take we should have bulk upsert query

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
